### PR TITLE
fix: use stability window for slot threshold

### LINKED ISSUE #591: Calculate blockfetch batch slot threshold from protocol params
We do bulk blockfetch operations until we get relatively close to chain tip, determined by a "slot threshold". We've got the slot threshold hard-coded, but it should probably be the same as the stability window

### PR DESCRIPTION
**Pull request by wolf31o2**
## Summary

This PR replaces the hard-coded slot threshold for bulk blockfetch operations with the protocol-defined stability window, addressing the TODO comment in the codebase.

### Core Changes
- **Replaced hard-coded ** with dynamic calculation based on the stability window
- **Era-aware security parameter selection**:
  - Byron era: Uses  parameter from 
  - Shelley+ eras: Uses  from 
- **Updated stability window calculation**:
  - Byron era: stability window =  (security parameter)
  - Shelley+ eras: stability window =  (where f is active slots coefficient)

### Files Modified
- : Updated blockfetch slot threshold logic and epoch nonce calculation
- : Updated block validation to use correct security parameters

## Benefits

1. **Protocol Compliance**: Now uses the proper Cardano protocol-defined security parameters
2. **Era Awareness**: Correctly handles different security parameter sources across Byron and Shelley+ eras  
3. **Network Flexibility**: Automatically adapts to different network configurations (mainnet, testnet, etc.)
4. **Maintainability**: Eliminates hard-coded values and TODO comments

## Testing

- [x] Code compiles successfully
- [x] All existing tests pass
- [x] No breaking changes to existing functionality

## Technical Details

The stability window is calculated differently based on the era:
- **Byron era (ID 0)**: Uses the  parameter directly as the stability window
- **Shelley+ eras**: Calculates  where  is the security parameter and  is the active slots coefficient from genesis

This ensures that bulk blockfetch operations stop at the appropriate distance from the chain tip, as defined by the protocol's stability guarantees.

## Summary by CodeRabbit

* **Bug Fixes**
  * Block-fetch gating and header handling now use an era-aware stability threshold with safe defaults, reducing premature catch-up exits and clearer warnings when genesis/config data is missing or invalid.

* **Refactor**
  * Epoch-nonce and stability-window logic consolidated into a single era-aware calculation with guarded behavior and fallbacks for missing or malformed genesis data.

* **Tests**
  * Extensive unit tests added covering stability-window and epoch-nonce scenarios and edge cases across all eras.

Closes #591

### CODE REVIEW

**Comment by wolf31o2** [comment]
@coderabbitai the tests for this are failing

File: ledger/state.go [hunk 1]
```diff
@@ -406,6 +406,96 @@ func (ls *LedgerState) consumeUtxo(
 	)
 }
 
+// calculateStabilityWindow returns the stability window based on the current era.
+// For Byron era, returns k. For Shelley+ eras, returns 3k/f.
+// Returns the default threshold if genesis data is unavailable or invalid.
+func (ls *LedgerState) calculateStabilityWindow() uint64 {
+	// Byron era only needs Byron genesis
+	if ls.currentEra.Id == 0 {
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply ledger/state.go]: In practice? Probably not.

File: ledger/state.go [hunk 2]
```diff
@@ -594,22 +684,42 @@ func (ls *LedgerState) ledgerProcessBlocks() {
...
+							k := shelleyGenesis.SecurityParam
+							if k < 0 {
+								return fmt.Errorf(
+									"security param must be non-negative: %d",
+									k,
+								)
+							}
+							// Safe conversion: SecurityParam is non-negative
+							securityParam = uint64(k) // #nosec G115
 						}
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply ledger/state.go]: I'm taking a simpler route

File: ledger/state.go [hunk 3]
```diff
@@ -406,6 +406,103 @@ func (ls *LedgerState) consumeUtxo(
...
+			ls.config.Logger.Warn("invalid negative security parameter", "k", k)
+			return blockfetchBatchSlotThresholdDefault
+		}
+		if k == 0 {
+			ls.config.Logger.Warn("security parameter is zero", "k", k)
+			return blockfetchBatchSlotThresholdDefault
+		}
+		// Byron uses k directly
+		return uint64(k) // #nosec G115
+	}
```
[the remark this answers is not in the document]
Comment (wolf31o2) [reply ledger/state.go]: Resolve the test failures.
